File: demo/js/frame_env.js
// Frame layer with physical-pose locking (PLAN2.md §1, amended).
//
// Bit-exact port of tetris/frame_env.py. The frame layer owns its board (v1
// row-int representation, row 0 top) and applies locks itself at the piece's
// TRUE PHYSICAL POSE — what the camera sees is what locks. Frozen v1 building
// blocks are reused read-only: rotation tables (via makeEngine's PIECES),
// Mulberry32 + SevenBag piece supply (identical queue discipline to the v1
// engine, so a seed yields the v1 piece sequence), and CLEAR_POINTS scoring.
//
// Semantics (frozen §1):
//   * 30 Hz ticks; gravity descends 1 row every GRAVITY_PERIOD (24) ticks.
//   * an action may be emitted every DECISION_PERIOD (3rd) tick (10 Hz).
//   * actions: noop / left / right / rot_cw / rot_ccw. A slide moves 1 column
//     if the destination cells are collision-free, else a silent no-op.
//     Rotation keeps the top-left anchor, clamps column to [0, 10 - width], and
//     fails (silent no-op) on collision. No kicks.
//   * spawn: rot 0, col floor((10 - width) / 2), bbox bottom at board row -1.
//   * lock: on a colliding gravity descent the piece locks at its current
//     physical pose: place cells, clear full rows shifting down, score via
//     CLEAR_POINTS, draw the next piece. Game over iff any locked cell has
//     row < 0 (board left unchanged, mirroring the v1 illegal-step outcome) or
//     the next spawn pose collides.
//
// Tucks: a slide can move a falling piece under an overhang so it rests deeper
// than the v1 straight drop of its (rot, col); the lock resolves at that
// physical pose and the lock event carries a `tuck` flag. v1-consistency
// invariant (fixture-tested): every non-tuck lock transition is bit-identical
// to v1 engine.step(r, c). Runs unmodified in the browser and Node.

import { makeEngine, CLEAR_POINTS } from "./engine.js";
import { WIDTH, HEIGHT, FULL_ROW } from "./features.js";
import { Mulberry32, SevenBag } from "./rng.js";

export const TICK_HZ = 30;
export const GRAVITY_PERIOD = 24;
export const DECISION_PERIOD = 3;

export const ACTIONS = ["noop", "left", "right", "rot_cw", "rot_ccw"];
export const NOOP = 0;
export const LEFT = 1;
export const RIGHT = 2;
export const ROT_CW = 3;
export const ROT_CCW = 4;

export function makeFrameEnv(piecesJson) {
  const { PIECES } = makeEngine(piecesJson);

  class FrameEnv {
    constructor(seed = 0, preview = 5) {
      this.reset(seed, preview);
    }

    reset(seed, preview = 5) {
      this.rng = new Mulberry32(seed);
      this.bag = new SevenBag(this.rng);
      this.preview = preview;
      this.rows = new Array(HEIGHT).fill(0);
      // Piece supply: identical queue discipline to the v1 engine.
      this.queue = [];
      this._fillQueue();
      this.piece = this.queue.shift();
      this._fillQueue();
      this.lines = 0;
      this.pieces = 0;
      this.score = 0; // sum of CLEAR_POINTS[lines] * 100, v1 demo convention
      this.tickCount = 0;
      this.gravityCounter = 0;
      this.gameOver = false;
      this._pending = NOOP;
      this._spawn();
      if (this._collides(this.piece, this.rot, this.col, this.row)) {
        this.gameOver = true;
      }
      return this;
    }

    _fillQueue() {
      while (this.queue.length < this.preview) this.queue.push(this.bag.nextPiece());
    }

    previewPieces() {
      return this.queue.slice(0, this.preview);
    }

    _spawn() {
      const rot0 = PIECES[this.piece][0];
      this.rot = 0;
      this.col = Math.floor((WIDTH - rot0.width) / 2);
      this.row = -rot0.height; // bounding-box bottom at board row -1
    }

    get isDecisionTick() {
      return this.tickCount % DECISION_PERIOD === 0;
    }

    pose() {
      return [this.piece, this.rot, this.col, this.row];
    }

    _collides(piece, rot, col, row) {
      const rows = this.rows;
      for (const [ro, co] of PIECES[piece][rot].cells) {
        const rr = row + ro;
        const cc = col + co;
        if (cc < 0 || cc >= WIDTH) return true;
        if (rr >= HEIGHT) return true;
        if (rr >= 0 && (rows[rr] >> cc) & 1) return true;
      }
      return false;
    }

    _straightDropRow(rotIdx, col) {
      // v1 straight-drop resting top-row for the active piece at (rot, col).
      const rot = PIECES[this.piece][rotIdx];
      let t = HEIGHT;
      for (let pc = 0; pc < rot.width; pc++) {
        const c = col + pc;
        let top = HEIGHT;
        for (let r = 0; r < HEIGHT; r++) {
          if ((this.rows[r] >> c) & 1) {
            top = r;
            break;
          }
        }
        const v = top - 1 - rot.bottom[pc];
        if (v < t) t = v;
      }
      return t;
    }

    apply_action(action) {
      if (!this.isDecisionTick) {
        throw new Error("apply_action is only valid on a decision tick");
      }
      this._pending = action;
    }

    _doAction(action) {
      if (action === NOOP) return;
      if (action === LEFT || action === RIGHT) {
        const nc = this.col + (action === LEFT ? -1 : 1);
        if (!this._collides(this.piece, this.rot, nc, this.row)) this.col = nc;
        return;
      }
      const n = PIECES[this.piece].length;
      const nr =
        action === ROT_CW ? (this.rot + 1) % n : (this.rot - 1 + n) % n;
      const w = PIECES[this.piece][nr].width;
      const nc = Math.min(Math.max(this.col, 0), WIDTH - w);
      if (!this._collides(this.piece, nr, nc, this.row)) {
        this.rot = nr;
        this.col = nc;
      }
    }

    tick() {
      if (this.gameOver) {
        this.tickCount += 1;
        return null;
      }

      if (this.tickCount % DECISION_PERIOD === 0) this._doAction(this._pending);
      this._pending = NOOP;

      let lock = null;
      this.gravityCounter += 1;
      if (this.gravityCounter >= GRAVITY_PERIOD) {
        this.gravityCounter = 0;
        if (this._collides(this.piece, this.rot, this.col, this.row + 1)) {
          lock = this._lockAndSpawn();
        } else {
          this.row += 1;
        }
      }

      this.tickCount += 1;
      return lock;
    }

    _lockAndSpawn() {
      const rotIdx = this.rot;
      const col = this.col;
      const row = this.row;
      const rot = PIECES[this.piece][rotIdx];

      // Straight drop is the highest reachable rest; deeper == tuck.
      const drop = this._straightDropRow(rotIdx, col);
      if (drop > row) {
        throw new Error(
          `straight-drop invariant violated: drop row ${drop} > lock row ` +
            `${row} (piece ${this.piece} rot ${rotIdx} col ${col})`,
        );
      }
      const tuck = drop !== row;

      // Any locked cell above the board => game over; board left unchanged
      // (bit-identical to the v1 illegal-step outcome). The bbox top row
      // always contains a cell, so row < 0 is the exact test.
      if (row < 0) {
        this.gameOver = true;
        return {
          tick: this.tickCount,
          r: rotIdx,
          c: col,
          lines_after: this.lines,
          tuck,
        };
      }

      // Physical lock: place cells, clear full rows shifting above down
      // (identical row surgery to v1 engine._apply).
      for (const [ro, co] of rot.cells) {
        this.rows[row + ro] |= 1 << (col + co);
      }
      let lines = 0;
      for (let r = row; r < row + rot.height; r++) {
        if (this.rows[r] === FULL_ROW) lines += 1;
      }
      if (lines) {
        const kept = this.rows.filter((v) => v !== FULL_ROW);
        this.rows = new Array(lines).fill(0).concat(kept);
      }
      this.lines += lines;
      this.score += CLEAR_POINTS[lines] * 100;
      this.pieces += 1;
      const lock = {
        tick: this.tickCount,
        r: rotIdx,
        c: col,
        lines_after: this.lines,
        tuck,
      };

      // Draw the next piece and spawn; a colliding spawn pose ends the game.
      this.piece = this.queue.shift();
      this._fillQueue();
      this._spawn();
      if (this._collides(this.piece, this.rot, this.col, this.row)) {
        this.gameOver = true;
      }
      return lock;
    }
  }

  return { FrameEnv, PIECES };
}

File: tetris/frame_env.py
"""Frame layer over the v1 atomic engine (PLAN2.md §1).

The v1 :class:`~tetris.engine.TetrisEngine` remains ground truth for lock,
line-clear, and game-over. This layer adds real-time semantics on top of it:

* Logic runs at 30 Hz (ticks).
* Gravity: the active piece descends 1 row every ``GRAVITY_PERIOD`` (24) ticks.
* Decisions: an action may be emitted every ``DECISION_PERIOD`` (3rd) tick
  (10 Hz "hand rate"); intermediate ticks advance gravity only.
* Actions: ``noop``, ``left``, ``right``, ``rot_cw``, ``rot_ccw`` — a slide
  moves 1 column if the destination cells are collision-free, else a silent
  no-op; rotation keeps the current top-left anchor with the column clamped to
  ``[0, 10 - width]`` and fails (silent no-op) on collision. No kicks.
* Spawn: rotation 0, column ``floor((10 - width) / 2)``, bounding-box bottom at
  board row -1 (fully above the board), on the tick after lock. A spawn pose
  that collides ends the game.
* Lock: when a gravity descent would collide, the piece locks at its current
  pose. The equivalent v1 placement is simply ``(rot, col)`` — under
  gravity-only descent (no soft drop, no reachable overhang tuck) the frame
  resting row is exactly the v1 straight-drop resting row. The frame layer
  calls ``engine.step(rot, col)`` so the board transition is v1-consistent by
  construction; an above-row-0 lock maps to an engine-illegal placement and
  ends the game.

Straight-drop invariant (and its one exception: tucks)
------------------------------------------------------
For the common case the frame resting row equals ``engine._drop_row(rot, col)``:
a piece descends straight down and rests exactly where a v1 straight drop would.

There is one reachable exception. Because a slide only checks the *destination*
cells for collisions (frozen §1 semantics) and there are 8 decision ticks per
gravity row, a piece falling down an open column can slide sideways *under* an
overhang into a covered cell — a "tuck". Straight-drop-only play does create
overhangs (S/Z/T pieces on uneven terrain), so tucks are genuinely reachable and
occur in ordinary random play. A tucked piece rests deeper than any v1 straight
drop of its final ``(rot, col)`` can reach, so the true, always-holding invariant
is ``engine._drop_row(rot, col) <= frame_row`` (straight drop is the *highest*
reachable rest); equality is the no-tuck case.

The frozen spec mandates locking via ``engine.step(rot, col)`` regardless, so the
board transition stays v1-consistent by construction in both cases — on a tuck
the engine locks the piece at the (higher) straight-drop row. The lock event
carries a ``tuck`` flag so this is observable. See ``tests/test_frame_env.py``
for the invariant test, a constructed tuck, and rotation/clamp/above-board edges.
"""

from __future__ import annotations

from tetris.engine import PIECES, TetrisEngine
from tetris.features import HEIGHT, WIDTH

# Frozen timing constants (PLAN2.md §1).
TICK_HZ = 30
GRAVITY_PERIOD = 24  # ticks per 1-row gravity descent
DECISION_PERIOD = 3  # a decision may be emitted every 3rd tick

# Action space (PLAN2.md §1). Index order is frozen.
ACTIONS = ("noop", "left", "right", "rot_cw", "rot_ccw")
NOOP, LEFT, RIGHT, ROT_CW, ROT_CCW = range(5)


class FrameEnv:
    """Real-time frame layer wrapping the v1 engine (see module docstring)."""

    def __init__(self, seed: int = 0, preview: int = 5):
        self.reset(seed, preview)

    # -- lifecycle -----------------------------------------------------------

    def reset(self, seed: int, preview: int = 5) -> "FrameEnv":
        self.engine = TetrisEngine(seed=seed, preview=preview)
        self.tick_count = 0
        self.gravity_counter = 0
        self.game_over = False
        self._pending = NOOP
        self.piece = self.engine.current
        self._spawn()
        # Spawn pose is fully above the board, so this is defensive only.
        if self._collides(self.piece, self.rot, self.col, self.row):
            self.game_over = True
        return self

    def _spawn(self) -> None:
        rot0 = PIECES[self.piece][0]
        self.rot = 0
        self.col = (WIDTH - rot0.width) // 2
        self.row = -rot0.height  # bounding-box bottom at board row -1

    # -- introspection -------------------------------------------------------

    @property
    def is_decision_tick(self) -> bool:
        return self.tick_count % DECISION_PERIOD == 0

    @property
    def rows(self) -> list[int]:
        return self.engine.rows

    @property
    def lines(self) -> int:
        return self.engine.lines

    @property
    def pieces(self) -> int:
        return self.engine.pieces

    def pose(self) -> tuple[int, int, int, int]:
        """Active piece as (piece_id, rot, col, row)."""
        return (self.piece, self.rot, self.col, self.row)

    # -- geometry ------------------------------------------------------------

    def _collides(self, piece: int, rot: int, col: int, row: int) -> bool:
        """True if placing `piece` at (rot, col, row) overlaps a wall, the
        floor, or a filled stack cell. Cells above the board (row < 0) are free."""
        rows = self.engine.rows
        for ro, co in PIECES[piece][rot].cells:
            rr = row + ro
            cc = col + co
            if cc < 0 or cc >= WIDTH:
                return True
            if rr >= HEIGHT:
                return True
            if rr >= 0 and (rows[rr] >> cc) & 1:
                return True
        return False

    # -- actions -------------------------------------------------------------

    def apply_action(self, action: int) -> None:
        """Queue an action for the current decision tick. Only valid on a
        decision tick; the queued action is consumed by the next ``tick()``."""
        if not self.is_decision_tick:
            raise ValueError("apply_action is only valid on a decision tick")
        self._pending = action

    def _do_action(self, action: int) -> None:
        if action == NOOP:
            return
        if action == LEFT or action == RIGHT:
            nc = self.col + (-1 if action == LEFT else 1)
            if not self._collides(self.piece, self.rot, nc, self.row):
                self.col = nc
            return
        # Rotation: keep the top-left anchor, clamp column to [0, 10 - width].
        n = len(PIECES[self.piece])
        nr = (self.rot + 1) % n if action == ROT_CW else (self.rot - 1) % n
        w = PIECES[self.piece][nr].width
        nc = min(max(self.col, 0), WIDTH - w)
        if not self._collides(self.piece, nr, nc, self.row):
            self.rot = nr
            self.col = nc

    # -- tick ----------------------------------------------------------------

    def tick(self) -> dict | None:
        """Advance one tick. Returns a lock-event dict on the tick a piece
        locks, else ``None``."""
        if self.game_over:
            self.tick_count += 1
            return None

        if self.tick_count % DECISION_PERIOD == 0:
            self._do_action(self._pending)
        self._pending = NOOP

        lock = None
        self.gravity_counter += 1
        if self.gravity_counter >= GRAVITY_PERIOD:
            self.gravity_counter = 0
            if self._collides(self.piece, self.rot, self.col, self.row + 1):
                lock = self._lock_and_spawn()
            else:
                self.row += 1

        self.tick_count += 1
        return lock

    def _lock_and_spawn(self) -> dict:
        rot_idx = self.rot
        col = self.col
        rot = PIECES[self.piece][rot_idx]
        # Straight-drop invariant: the v1 straight-drop row is the highest
        # reachable rest, so it never sits below the frame resting row. Equality
        # is the no-tuck case; drop < row means the piece tucked under an
        # overhang (see module docstring). The lock always defers to
        # engine.step(rot, col) so the board transition stays v1-consistent.
        drop = self.engine._drop_row(rot, col, self.engine._col_top())
        assert drop <= self.row, (
            f"straight-drop invariant violated: engine drop row {drop} > frame "
            f"row {self.row} (piece {self.piece} rot {rot_idx} col {col})"
        )

        self.engine.step(rot_idx, col)
        lock = {
            "tick": self.tick_count,
            "r": rot_idx,
            "c": col,
            "lines_after": self.engine.lines,
            "tuck": drop != self.row,
        }
        if self.engine.game_over:
            self.game_over = True
            return lock

        self.piece = self.engine.current
        self._spawn()
        if self._collides(self.piece, self.rot, self.col, self.row):
            self.game_over = True
        return lock
